services/base.js

```diff
@@ -321,20 +319,28 @@ class BaseService {
     }
   }
 
-  async invokeHandler(namedParams, queryParams) {
-    trace.logTrace(
-      'inbound',
-      emojic.womanCook,
-      'Service class',
-      this.constructor.name
-    )
+  static async invoke(
+    context = {},
+    config = {},
+    namedParams = {},
+    queryParams = {}
+  ) {
```

Comment:
This is probably a slightly subjective point, but I'd be inclined to leave most of this method in an `invokeHandler()` method so it is easier to write something like:

```js
const service = new Service()
service.property = 'value'
service.invokeHandler()
```

If we retain `invokeHandler()` as a method and keep `invoke()` thin so its implementation is like:


```js
static async invoke(
  context = {},
  config = {},
  namedParams = {},
  queryParams = {}
) {
  const serviceInstance = new this(context, config)
  return await serviceInstance.invokeHandler(namedParams, queryParams)
}
```

That way we retain that flexibility if we need it, but you can still call `await Service.invoke()` in the majority of cases where you don't need to do that. I can see in the cases where we need to mock something you've handled it by extending the class to override a method, but I think keeping `invoke()` thin is probably a useful property. What do you think?

Reply:
Yea… agreed, subjective.

I'm not sure where outside of tests on BaseService subclasses we'd need to set these properties on classes, and there the subclass approach is working fine. After all, these are base classes, intended for subclassing.

I was trying to avoid "lasagne code." If spaghetti code is code where the wires are crossed all over the place and you can't follow what's happening, then lasagne code has lots of layers you have to peel through, none of which do much at all.

There was also the matter of naming: `invoke` and `invokeHandler` are basically two names that mean the same thing. I suppose we could name them _both_ invoke, since they are in different namespaces. Or name the class method something like `run` and leave the instance method as `invokeHandler`.

I certainly don't feel strongly about this.

Reply:
> I'm not sure where outside of tests on BaseService subclasses we'd need to set these properties

I'm mainly thinking of tests. I guess we can always extract this logic at a later date if we need to.